core/src/main/scala/kafka/log/Log.scala

```diff
@@ -364,10 +441,10 @@ class Log(@volatile var dir: File,
    * @return Information about the appended messages including the first and last offset.
    */
   def append(records: MemoryRecords, assignOffsets: Boolean = true): LogAppendInfo = {
-    val appendInfo = analyzeAndValidateRecords(records)
+    val appendInfo = analyzeAndValidateRecords(records, isFromClient = assignOffsets)
```

Comment:
In ReplicaFetcherThread, we probably want to log a warning if LogAppendInfo.isDuplicate is true after the append() call since it's not expected.